tools/ddprobe.cpp

```diff
@@ -69,6 +71,123 @@ syncThreadDesktop() {
+ * If all 10 frames are empty, it returns `S_FALSE`, suggesting potential issues with the capture process.
+ * If any error occurs during the frame acquisition or analysis process, the corresponding `HRESULT` error code is returned.
+ *
+ * @param dup A reference to the DXGI output duplication object (`dxgi::dup_t&`) used to acquire frames.
+ * @param device A ComPtr to the ID3D11Device interface representing the device associated with the Direct3D context.
+ * @return HRESULT Returns `S_OK` if a non-empty frame is captured successfully, `S_FALSE` if all frames are empty, or an error code if any failure occurs during the process.
+ */
+HRESULT
+test_frame_capture(dxgi::dup_t &dup, ComPtr<ID3D11Device> device) {
+  for (int i = 0; i < 10; ++i) {
+    std::cout << "Attempting to acquire frame " << (i + 1) << " of 10..." << std::endl;
+
+    ComPtr<IDXGIResource> frameResource;
+    DXGI_OUTDUPL_FRAME_INFO frameInfo;
+    HRESULT status = dup->AcquireNextFrame(500, &frameInfo, &frameResource);
+    auto release_frame = util::fail_guard([&dup]() {
+      dup->ReleaseFrame();
+    });
+
+    if (FAILED(status)) {
+      std::cout << "Error: Failed to acquire next frame [0x"sv << util::hex(status).to_string_view() << ']' << std::endl;
+      return status;
+    }
+
+    std::cout << "Frame acquired successfully." << std::endl;
+
+    ComPtr<ID3D11Texture2D> frameTexture;
+    status = frameResource->QueryInterface(__uuidof(ID3D11Texture2D), reinterpret_cast<void **>(frameTexture.GetAddressOf()));
+    if (FAILED(status)) {
+      std::cout << "Error: Failed to query texture interface from frame resource [0x"sv << util::hex(status).to_string_view() << ']' << std::endl;
+      return status;
+    }
+
+    D3D11_TEXTURE2D_DESC frameDesc;
+    frameTexture->GetDesc(&frameDesc);
+    frameDesc.Usage = D3D11_USAGE_STAGING;
+    frameDesc.CPUAccessFlags = D3D11_CPU_ACCESS_READ;
+    frameDesc.BindFlags = 0;
+    frameDesc.MiscFlags = 0;
+
+    ComPtr<ID3D11Texture2D> stagingTexture;
+    status = device->CreateTexture2D(&frameDesc, nullptr, &stagingTexture);
+    if (FAILED(status)) {
+      std::cout << "Error: Failed to create staging texture [0x"sv << util::hex(status).to_string_view() << ']' << std::endl;
+      return status;
+    }
+
+    ComPtr<ID3D11DeviceContext> context;
+    device->GetImmediateContext(&context);
+    context->CopyResource(stagingTexture.Get(), frameTexture.Get());
+
+    D3D11_MAPPED_SUBRESOURCE mappedResource;
+    status = context->Map(stagingTexture.Get(), 0, D3D11_MAP_READ, 0, &mappedResource);
+    if (FAILED(status)) {
+      std::cout << "Error: Failed to map the staging texture for inspection [0x"sv << util::hex(status).to_string_view() << ']' << std::endl;
+      return status;
+    }
+
+    if (is_valid_frame(mappedResource, frameDesc)) {
+      std::cout << "Frame " << (i + 1) << " is non-empty (contains visible content)." << std::endl;
+      context->Unmap(stagingTexture.Get(), 0);
```

Comment:
`context->Unmap` could also be put into a guard.

Reply:
I was thinking about that too, but thought it might be overkill since it only saved one line of code… 

but since I already had a guard there, it made more sense to do that so good suggestion